### PR TITLE
feat(input): add utilities for custom styling and monitoring state of input autofill

### PR DESCRIPTION
**Pull request by mmalerba**
Demo: https://mmalerba-demo5.firebaseapp.com/input

### CODE REVIEW

**Comment by mmalerba** [comment]
@jelbourn moved everything into `MatInputModule` PTAL.

File: src/lib/input/autofill.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Directive, ElementRef, EventEmitter, Injectable, OnDestroy, Output} from '@angular/core';
+import {Observable} from 'rxjs/Observable';
+import {Subject} from 'rxjs/Subject';
+
+
+/** An event that is emitted when the autofill state of an input changes. */
+export type AutofillEvent = {
```
Comment (jelbourn) [inline src/lib/input/autofill.ts]: Make everything in here `@docs-private`?

File: src/lib/input/autofill.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      if (event.animationName === 'mat-input-autofill-start') {
+        element.classList.add('mat-input-autofilled');
+        result.next({target: event.target as Element, isAutofilled: true});
+      } else if (event.animationName === 'mat-input-autofill-end') {
+        element.classList.remove('mat-input-autofilled');
+        result.next({target: event.target as Element, isAutofilled: false});
+      }
+    };
+
+    element.addEventListener('animationstart', listener);
```
Comment (jelbourn) [inline src/lib/input/autofill.ts]: Make this a passive listener and run outside the zone?

File: src/lib/input/autofill.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ * An injectable service that can be used to monitor the autofill state of an input.
+ * Based on the following blog post:
+ * https://medium.com/@brunn/detecting-autofilled-fields-in-javascript-aed598d25da7
+ */
+@Injectable()
+export class AutofillMonitor implements OnDestroy {
+  private _monitoredElements = new Map<Element, MonitoredElementInfo>();
+
+  monitor(element: Element): Observable<AutofillEvent> {
+    const info = this._monitoredElements.get(element);
```
Comment (jelbourn) [inline src/lib/input/autofill.ts]: No-op when not on the browser platform?

File: src/lib/input/autofill.ts (hunk 1 above)
Comment (mmalerba) [reply src/lib/input/autofill.ts]: Why? I do want people to be able to use these

Comment (jelbourn) [reply src/lib/input/autofill.ts]: Ack, I had thought there was going to be more iteration after this PR.

Comment (mmalerba) [reply src/lib/input/autofill.ts]: did add some additional doc comments though

**Comment by mmalerba** [comment]
comments addressed

**Pull request merged**